### PR TITLE
retry reconcile task until no platform errors

### PR DESCRIPTION
**Pull request by chengjoey**
#### What type of this PR

Add one of the following kinds:
/kind bug

#### What this PR does / why we need it:
add two error type(platform, user)
retry reconcile task always  if error is platform error

#### Which issue(s) this PR fixes:

- Fixes #your-issue_number
- [Erda Cloud Issue Link](https://erda-org.erda.cloud/erda/dop/projects/387/issues/all?id=207028&issueFilter__urlQuery=eyJzdGF0ZUJlbG9uZ3MiOlsiT1BFTiIsIldPUktJTkciLCJXT05URklYIiwiUkVPUEVOIl0sImFzc2lnbmVlSURzIjpbIjEwMDEyMDUiXX0%3D&issueTable__urlQuery=eyJwYWdlTm8iOjF9&issueViewGroup__urlQuery=eyJ2YWx1ZSI6ImthbmJhbiIsImNoaWxkcmVuVmFsdWUiOnsia2FuYmFuIjoiZGVhZGxpbmUifX0%3D&iterationID=506&type=TASK)

#### Specified Reviewers:

/assign @your-reviewer
@sfwn

### CODE REVIEW

File: modules/pipeline/pipengine/reconciler/task.go [hunk 1]
```diff
@@ -83,24 +84,32 @@ func reconcileTask(tr *taskrun.TaskRun) error {
 			// Do 没有 err 才是正常的，即使失败也是 status=Failed，没有 error
 			abnormalErr := tr.Do(taskOp)
 			if abnormalErr != nil {
-				rlog.TWarnf(tr.P.ID, tr.Task.ID, "failed to handle taskOp: %s, abnormalErr: %v, continue retry", taskOp.Op(), abnormalErr)
+				if errors.IsContainUserError(abnormalErr) {
+					rlog.TErrorf(tr.P.ID, tr.Task.ID, "failed to handle taskOp: %s, user abnormalErr: %v, don`t need retry", taskOp.Op(), abnormalErr)
```
Comment (sfwn) [inline modules/pipeline/pipengine/reconciler/task.go:88]: don't

File: modules/pipeline/pipengine/reconciler/task.go [hunk 2]
```diff
@@ -83,24 +84,32 @@ func reconcileTask(tr *taskrun.TaskRun) error {
...
+					return abnormalErr
+				}
+				// don`t contain user error mean err is platform error, should retry always
```
Comment (sfwn) [inline modules/pipeline/pipengine/reconciler/task.go:91]: don't

File: modules/pipeline/pipengine/reconciler/task.go [hunk 3]
```diff
@@ -83,24 +84,32 @@ func reconcileTask(tr *taskrun.TaskRun) error {
...
+				rlog.TWarnf(tr.P.ID, tr.Task.ID, "failed to handle taskOp: %s, abnormalErr: %v, continue retry, retry times: %d", taskOp.Op(), abnormalErr, platformErrRetryTimes)
```
Comment (sfwn) [inline modules/pipeline/pipengine/reconciler/task.go:92]: Errorf

File: modules/pipeline/pipengine/reconciler/task.go [hunk 4]
```diff
@@ -83,24 +84,32 @@ func reconcileTask(tr *taskrun.TaskRun) error {
...
-				// 小于异常重试次数，继续重试
-				if platformErrRetryTimes < abnormalErrMaxRetryTimes {
-					resetTaskForAbnormalRetry(tr, platformErrRetryTimes)
-					platformErrRetryTimes++
-					continue
-				}
-				// 大于重试次数仍有异常，返回异常
-				rlog.TErrorf(tr.P.ID, tr.Task.ID, "failed to handle taskOp: %s, abnormalErr: %v, reach max retry times, return abnormalErr", taskOp.Op(), abnormalErr)
-				return abnormalErr
+				//if tr.IsSessionNotFound(abnormalErr) {
```
Comment (sfwn) [inline modules/pipeline/pipengine/reconciler/task.go:98]: delete useless codes

File: modules/pipeline/pkg/errors/errors.go [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+// or later ("AGPL"), as published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful, but WITHOUT
+// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+// FITNESS FOR A PARTICULAR PURPOSE.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+package errors
```
Comment (sfwn) [inline modules/pipeline/pkg/errors/errors.go:14]: errorsx

File: modules/pipeline/pipengine/reconciler/task.go [hunk 6]
```diff
@@ -83,24 +84,32 @@ func reconcileTask(tr *taskrun.TaskRun) error {
 			// Do 没有 err 才是正常的，即使失败也是 status=Failed，没有 error
 			abnormalErr := tr.Do(taskOp)
 			if abnormalErr != nil {
-				rlog.TWarnf(tr.P.ID, tr.Task.ID, "failed to handle taskOp: %s, abnormalErr: %v, continue retry", taskOp.Op(), abnormalErr)
+				if errors.IsContainUserError(abnormalErr) {
```
Comment (sfwn) [inline modules/pipeline/pipengine/reconciler/task.go:87]: Can you use IsUserError directly?

**Review by sfwn — CHANGES_REQUESTED** [review]

**Comment by sfwn** [comment]
Add a test log screenshot.

File: modules/pipeline/pipengine/reconciler/task.go (hunk 6 above)
Comment (chengjoey) [reply modules/pipeline/pipengine/reconciler/task.go:87]: image
task.waitOp aggregate the errors list to one error, the aggregate error may contain platform error and user error

**Comment by chengjoey** [comment]
continue retry if platform error, and aggregate same errors
image

File: modules/pipeline/pipengine/reconciler/taskrun/framework.go [hunk 7]
```diff
@@ -151,7 +167,8 @@ func (tr *TaskRun) waitOp(itr TaskOp, o *Elem) (result error) {
 	case err := <-o.ErrCh:
 		logrus.Errorf("reconciler: pipelineID: %d, task %q %s received error (%v)", tr.P.ID, tr.Task.Name, itr.Op(), err)
 		if tr.IsSessionNotFound(err) {
-			errs = append(errs, "Failed to find Session")
+			// convert session error
+			errs = append(errs, fmt.Sprintf("Cluster: %s network abnormal", tr.Task.Extra.ClusterName))
```
Comment (sfwn) [inline modules/pipeline/pipengine/reconciler/taskrun/framework.go:171]: Network issue for cluster: %s (%s)           clusterName, reason like timeout

Comment (sfwn) [reply modules/pipeline/pipengine/reconciler/taskrun/framework.go:171]: maybe other network issue, distinguish is meta is better.

File: modules/pipeline/services/pipelinesvc/callback.go [hunk 8]
```diff
@@ -80,8 +80,14 @@ func (s *PipelineSvc) appendPipelineTaskResult(p *spec.Pipeline, task *spec.Pipe
 	}
 	// metadata
 	task.Result.Metadata = append(task.Result.Metadata, cb.Metadata...)
-	// error
-	task.Result.Errors = append(task.Result.Errors, cb.Errors...)
+	// TODO action agent should add err start time and end time
+	taskErrors := make([]*apistructs.PipelineTaskErrResponse, 0)
```
Comment (sfwn) [inline modules/pipeline/services/pipelinesvc/callback.go:84]: newTaskErrors

Comment (sfwn) [reply modules/pipeline/services/pipelinesvc/callback.go:84]: add unit test.

**Review by sfwn — CHANGES_REQUESTED** [review]

File: modules/pipeline/pipengine/reconciler/taskrun/framework.go [hunk 9]
```diff
@@ -150,8 +161,9 @@ func (tr *TaskRun) waitOp(itr TaskOp, o *Elem) (result error) {
 
 	case err := <-o.ErrCh:
 		logrus.Errorf("reconciler: pipelineID: %d, task %q %s received error (%v)", tr.P.ID, tr.Task.Name, itr.Op(), err)
-		if tr.IsSessionNotFound(err) {
-			errs = append(errs, "Failed to find Session")
+		if errorsx.IsNetworkError(err) {
+			// convert network error
+			errs = append(errs, fmt.Sprintf("Network issue for cluster: %s (%v) clusterName, reason like timeout", tr.Task.Extra.ClusterName, err))
```
Comment (sfwn) [inline modules/pipeline/pipengine/reconciler/taskrun/framework.go:166]: remove 'clusterName, reason like timeout', just a sample for %s %s

**Review by sfwn — CHANGES_REQUESTED** [review]

**Review by sfwn — APPROVED** [review]

**Pull request merged**